Prototype for Cygwin support module (not ready to merge)

embray opened the pull request:
Here is my prototype implementation for a Cygwin support module for psutil (#82).

This has been tested on Python 2.7 and 3.4 and all the tests pass on those versions (Pythons >= 3.5 are not supported on Cygwin yet).  There are a few tests that are skipped on Cygwin--mostly for corner cases that don't work on Cygwin the way they do, say, on Linux.

The main TODO items as I see it in order of effort / importance are:

* [ ] Refactoring--there are large chunks of code copied from the Linux and Windows modules to support different features on Cygwin.  The copied code has either no, or only minor modifications.  I'm thinking of moving such code around so that it can be shared by all modules that use it (to be clear, the full Windows module does *not* work on Cygwin and shouldn't be compiled there, but bits of it do; likewise for the Linux module).
* [ ] Documentation needs to be updated to reflect the new platform.
* [ ] Fix testing on Python 3.4 for Cygwin on AppVeyor
* [ ] There are still a number of compilation warnings that should probably be cleaned up if possible.

Commit:
Ignore dlls from running build_ext --inplace on Windows

Commit:
Initial addition of a cygwin platform module.

Some common utilitiy functions were moved out of _pslinux into _common.

Most of cygwin's Process implementation is directly copied from the
linux Proces, minus a few very Linux-specific parts.  I hope to
eliminate the code duplication later with further refactoring.

Most tests fail due to some key Process methods not having
implementations yet (mostly because their implementations for Linux do
not work on Cygwin).

Commit:
First pass at implementing Process.create_time for Cygwin

I left this unimplemented originally just because I wanted to take the
time to understand what I was doing before continuing.  Turns out it can
be implemented almost the same as on Linux, except that the process
start time is in the 22nd field of /proc/<pid>/stat instead of the 21st.
I can't find enough documentation for Cygwin's /proc filesystem, but I
confirmed this in the source.  Other than that the semantics are the
same as on Linux.

So much of this was copied agained from the Linux module, but I will
refactor later.

Commit:
Add pids() and pid_exists(), again copied verbatim from the Linux
module.

Commit:
Added initial virtual_memory() and swap_memory() implementations.

These are much simpler than on Linux, and effectively the same as the
existing Windows implementations of these functions, though easier to
implement on Cygwin via /proc/meminfo

Commit:
Add previously missing pfullmem namedtuple, which some of the tests
expect to find.

Currently it is identical to pmem; will return later to seeing what
additional memory info we can provide on Cygwin

Commit:
Fix broken Process.uids and Process.gids on Cygwin

Later this should be able to share its implementation with the Linux
module but we'll come back around to that.

Commit:
Add cpu_count_logical(), cpu_count_physical(), and cpu_stats() for
Cygwin

Copied almost verbatim from the Linux module.

Commit:
Compile at least the arch/windows/process_info.c module into the
cygwin module so that the functions in it can be reused there.

This required some tweaks to ntextapi.h, since a few of the
types defined there are already included in the winimpl.h included
with Cygwin.

This includes are own implementation of PyErr_SetFromWindowsErr which
is not defined in Python for Cygwin.

TODO: Because this PyErr_SetFromWindowsErr is lifted, only slightly
modified, from the Python source it should include a copyright notice
to the effect.  This implementation probably does not work quite right
on Python 2 either--it should return an actual WindowsError there.

Commit:
Add proc_cpu_affinity_get/set.  This is copied exactly from the Windows
module for now.

Will refactor later.

Commit:
Added an important TODO note on exception handling; will come back to this later.

Commit:
Initial disk_partitions() implementation for cygwin module.

Copied exactly from the Linux module and seems to work; will refactor
later.

Commit:
Initial implementation of disk_usage() for cygwin.

Using the posix implementation directly seems to just work.

Commit:
Initial implementation of net_if_addrs() for the cygwin module.

The only obvious issue currently is that rather than the human-friendly interface
names returned by this function in the normal Windows implementation, the
interface names are instead some UUID looking things for each interface (the same
as appear in the registry in various places to represent the interface).  This is
fine for now, but not ideal.  Should be able to map these to friendlier names.

Commit:
Add some previously missing declarations

Commit:
Move the py2_strencode utility from the windows module into common,
renaming it to 'encode', for symmetry with the 'decode' utility.

This utility may also be useful in the Cygwin module, and is otherwise
not Windows-specific, having more to do with Py2/3 compatibility.

Commit:
Very rough first pass at implementing net_if_stats for Cygwin

This is mostly copied from the Windows module with a few small tweaks, but compiling this
is a bit of a pain and somewhat fragile.  Will have to come back to this for a closer look
later, but for now I just want to get all of the API implemented as best I can.

Commit:
Initial implementation of net_io_counters for the cygwin module.

As with net_if_stats, this is a bit of a mess right now to compile, and copies from the Windows module, but will clean up later.

Commit:
Initial implementation of the net_connections() function for cygwin

Once again, mostly copied from the Windows module, with much refactoring needed.

Commit:
Added a simple wrapper for cygwin_conv_path() from Cygwin's (very limited) API.

This will allow methods that return paths to return Cygwin paths instead of native Windows paths.
I have yet to go through and determine all the places this should be used, but there are probably
at least a few.

Commit:
Add conversion from Cygwin PID to native Windows PID.

Whenever calling a C extension method that calls out to the Windows API make sure to use
the Windows PID.  This should fix a lot of issues where the wrong PID was being used.

Commit:
Initial implementation of Process.memory_maps for the cygwin module.

Most of this is copied from the windows module with a few minor tweaks, necessitating later
refactoring.

This is also the first method to make proper use of the new winpath_to_cygpath function.

Commit:
Fix cygwin's Process.cpu_affinity_get/set by copying the Python-level implementations from the windows module (rather than the linux module) but using the winpid.

Commit:
Initial implementation of disk_io_counters for Cygwin; copied almost verbatim from the windows module.

Commit:
Initial implementation of the psutil.users() function for the cygwin module.

This is copied mostly from the Linux module, though it needs to use the utmpx APIs.  This seems to 'work'
insofar as the API is available on cygwin, but it's not clear to me that the information returned is fully
correct or useful.  Will return to this if needed.

Commit:
On cygwin (as on windows) Process.connections is just a wrapper around net_connections.

Commit:
Added initial support for Process.environ, again copied from the windows module

(The linux module's method of looking in /proc doesn't work on cygwin's more limited /proc namespace)

Commit:
Initial cygwin support for Process.io_counters and Process.num_ctx_switches, mostly
supported by the proc_info function copied from the windows module.

Commit:
Add cygwin interface for Process.num_threads

(This is just a thin wrapper around proc_info())

Commit:
Should pass self._winpid to proc_info, not self.pid

Commit:
Initial cygwin support for Process.threads

Copy/pasted from windows module, so this carries the usual admonition about refactoring.

Commit:
The child process search that occurs if ppid_map is not available should fail gracefully on AccessDenied errors.

If we get an AccessDenied trying to get the ppid of a process, it's *probably* not a child process of the process in
question anyways.

Commit:
Use the Process.memory_info implementation from the windows module
instead of going through Cygwin's /proc which turns out to be a bit buggy and
for this purpose.

This allows us to expose more information anyways such as USS in memory_info_full,
just unfortunate that the Cygwin interface can't be relied on here.

Commit:
Turns out there does not (for now) appear to be any straightforward way
to get the process environment on Cygwin.  So we remove the
functionality for now.

See https://cygwin.com/ml/cygwin/2016-11/msg00205.html

Commit:
After researching the subject extensively, I've determined it's not possible to enumerate UNIX sockets on Cygwin

Cygwin emulates UNIX sockets as a INET socket.  Although it does list them in /proc/<pid>/fd as some 'socket:[N]' like on Linux, the 'N' here is only of internal meaning to Cygwin (this is true on Linux as well), but there is nothing resembling /proc/net/unix or anything else that allows one to map that number N to an actual socket file (such a file *does* exist, and in theory we could find it be enumerating ALL files on the system until it is found, but this is not practical).  I can't find any other obvious way to do that.  The only practical solution would be if Cygwin had an equivalent to /proc/net/unix.

Commit:
Add Cygwin version info constants--these are not used anywhere yet but may prove useful later

Commit:
allow process_iter to skip processes on AccessDenied, which can happen when trying to query processes for ppid that one does not have access to (TODO: it may be possible to get the ppid by some other means in this case though)

Commit:
Add a winpid_to_cygpid utility. Improve error handling in cygpid_to_winpid.

Commit:
Fix net_connections to return Cygwin PIDs instead of Windows PIDs

Commit:
ERROR_ACCESS_DENIED handling in a function where it's been coming up

Commit:
Skip a few tests that cannot work on Cygwin

Commit:
Catch AccessDenied errors in some tests that loop over all processes

Commit:
Adds a new ps() interface for the posix tests that better clarifies the
platform-specific differences between how ps is called.

Also implements significant support for Cygwin, where the built-in ps
lacks many of the features of ps on other systems.

This fixes a number of test errors on Cygwin (it also exposes a few new
failures but those will be addressed separately).

Commit:
Cygwin does not necessarily create utmp entries when you start a login terminal, so
'who' can return empty--this fix allows the test to work even when there are no utmp entries.

Commit:
Fix a corner case that can occur when a getmntent() entry's mount point
does not actually exist.

This happened to be affecting my test runs.

Commit:
Turns out the process creation time from /proc/<pid>/stat cannot be
relied on in Cygwin, due to a rounding error there which can cause it to
return fluctuating results over time.

This is really bad because process creation time is part of the
'identity' psutil util uses for a process when comparing it for equality
to another process.

This once again copies code from the windows module into the cygwin
module, which I intend to refactor later.  I also plan to address the
upstream bug in Cygwin, but until that's fixed this workaround is
needed.

Commit:
A TODO note for later

Commit:
It seems this test is having the same problem on Cygwin as on plain
Windows with processes not being truly destroyed quickly enough.

Commit:
Make one small Cygwin-specific ifdef in psutil_net_if_addrs to support getting the
'friendly name' as opposed to the fixed UUID-based name of an interface.  This is more consistent
with the other network APIs on Windows/Cygwin which return the interface friendly names

Commit:
Change net_if_stats() to return the same network interfaces as net_if_addrs()--kind of a hack for now.

Commit:
Make this test pass more reliably by allowing the ps output to include one extra process for ps itself.

(I'm not sure how this test worked on Linux since it normally includes ps itself too...)

Commit:
As written, this test could not work when usernames contain spaces (which is not uncommon on Cygwin)

Commit:
Prevent this test from failing on Cygwin on processes running at a higher privilege than the test runner

Commit:
memory_info is one area where the Cygwin module is more like Windows than POSIX

Commit:
These tests are a mess on Python 2 due to inconsitent handling of non-unicode strings

Commit:
Skip this test on Cygwin, as there is no reliable way currently to map from net_connections() results to file descriptors

Commit:
Improved handling for a buggy (?) state that can occur on Cygwin where a process is in the zombie state but its /proc/<pid>/stat file can no longer be read.

See https://www.cygwin.com/ml/cygwin/2017-02/msg00187.html

Commit:
Skip this portion of this test on Cygwin due to the reasons noted in the comments

Commit:
Return fewer interfaces from net_io_counters on Cygwin, for parity with the functionality in the Windows interface.

This allows the test_nic_names test from the Windows tests to be reused
and to pass on Cygwin.

Commit:
More fixes for this test to work on Cygwin.

I'm worried it still might be a bit flakey though.

Commit:
Fix Process.terminal() for ptys on Cygwin

Commit:
Sometimes Cygwin will show disk partitions whose mount points no longer exist

Commit:
Missing compat import for Python 3

Commit:
some of the exe tests pass most reliably on Cygwin if we always drop the .exe right here

Commit:
Added oneshot support on Cygwin

Commit:
Add support for sensors_battery on Cygwin

(Copied directly from the Windows module for now)

Commit:
On Cygwin, ps shows real paths, while psutil.cmdline() will return symlinks

Commit:
Skip these tests on AppVeyor (under Cygwin only)

They are failing persistently due to fork errors.  No matter how I rebase the affected DLLs it seems
to have problems on these. I have given up trying to find out why for now.

Commit:
Support /dev/consN on Cygwin

Commit:
no need for verbose output from rebaseall now that it seems to be working

Commit:
This test still fails due to the odd way Cygwin handles /dev/console

Commit:
Skip this superfluous Process instantiation

Commit:
Work around a test failure with set CPU affinity that is occurring on AppVeyor

Commit:
More fork errors on Cygwin, now immediately upon running setup.py build (this did not happen before).

Try a rebaseall immediately before setup.py build.

Commit:
Need to add CYG_ROOT/bin to PATH or else Python 3 crashes for some reason

Commit:
Strip .exe explicitly

(previously this would mistakenly convert something like 'python2.7' to just 'python2'--if the latter
is a symlink this can break things)

Commit:
Retry get_test_subprocess

It seems that random fork errors can occur here on Cygwin, but not *always*. Retrying usually works.

Commit:
Disable testing Python 3 on Cygwin in AppVeyor for now until I can spend
more time figuring out what the problem is.

Commit:
Initial attempt to enable testing against Cygwin on AppVeyor

giampaolo commented:
This is mirable work, thanks a lot for doing this. As for refactoring, to me it looks like _pscygwin.py has more code in common with _pslinux.py rather than _pswindows.py.
As such, what I would recommend is to (whenever possible) keep Linux and Cygwin logic in _pslinux.py and use `if CYGWIN: ...` where necessary and when the logic differs too much just put it in _pscygwin.py. Something like this:

```
# _pscygwin.py

from . import _pslinux

# same logic
pids = _pslinux.pids

# new logic
def disk_io_counters():
      ...  
```

The trick is to be cautious about code which is executed at import time in _pslinux.py. Again, you can just use `if not CYGWIN: ...`, e.g.:

```
# _pslinux.py

if LINUX:
     PAGESIZE = os.sysconf("SC_PAGE_SIZE")
```

As for the C extension module, things may be more hairy. I took a look at your _psutil_cygwin.c and it seems it was a copy & paste of _psutil_windows.c with a lot of subsequent chances. I see a lot of differences but it looks like you moved a lot of stuff around (for no good reason?) but the main bits are pretty much the same, so I can't properly comment about that and would like to hear your input about it. 
Is it feasible to use the original _psutil_windows.c and use `#ifdef __CYGWIN__` where necessary? 

Last: I would like to try this but again, I have basically zero knowledge about cygwin. Can you provide some brief instructions on how to set this up on Linux so that I can play with it?

giampaolo commented:
Also, one question which may sound naive, but what target of users is this for? E.g. what is your use case?
I understand that in order to use psutil on cygwin you'll have to re/compile psutil from sources.
This also poses questions on whether we should provide exe / wheels binaries specific for cygwin, and that's something I'm not sure I want to do.

embray commented:
One comment I will make is that a lot more of the code duplication is actually in the C module, rather than the Python module.

embray commented:
> The trick is to be cautious about code which is executed at import time in _pslinux.py

That's the main thing I would worry about with this, and it could be annoying and/or fragile.  I think it makes (maybe) more sense to move common functionality into a common module.  Something like a `_pslinux_common.py` or `_pslinux_utils.py`.  But maybe you prefer not to litter functionality across too many individual files.

embray commented:
Looks like I've still got some work to do on the tests too.  I'll come back to that.

embray commented:
> Is it feasible to use the original _psutil_windows.c and use #ifdef __CYGWIN__ where necessary?

Yes, I think that might be possible.  That said, there are a number of bits in the `_psutil_windows.c` module that *won't* compile under Cygwin.  That suggests an argument (similar to the one I gave above) for a `_psutil_windows_common.c` or something.

As for the functions I copied + modified, I'd have to take a look at them again (a lot of this I did over several months, so I don't remember all the changes), but I think in most cases a single version of the function (perhaps with an additional argument added, or some ifdefs) can serve both cases.

embray commented:
> I would like to try this but again, I have basically zero knowledge about cygwin. Can you provide some brief instructions on how to set this up so that I can play with it?

I'll get back to you on that.  I'd need to come up with a minimal set of instructions.  But for the most part there's not much to it.

embray commented:
> Also, one question which may sound naive, but what target of users is this for? E.g. what is your use case?
I understand that in order to use psutil on cygwin you'll have to re/compile psutil from sources.
This also poses questions on whether we should provide exe / wheels binaries specific for cygwin, and that's something I'm not sure I want to do.

This is *originally* motivated by SageMath, which recently started using `psutil`, and relies on Cygwin for Windows support.  That said, I personally just thought it would be an interesting exercise.  I don't think you need to worry about supplying wheels; I would probably make a python-psutil package for Cygwin eventually anyways.

Commit:
The 'Version' string in the output of 'ver' appears to be locale-dependent

embray commented:
Been busy with other things but I'm still planning to come back to this and do some more cleanup.

Commit:
Better way to get the major/minor version of Windows, without relying on running a subprocess
or cmd.exe.

Doesn't return the build number unfortunately, but we're not using that right now anyways.

embray commented:
Sorry for leaving this open for so long; I'm going to try to work some more on this now, to finally finish up the various loose ends.

Commit:
Workaround needed for a bug in Cygwin

giampaolo commented:
No rush. I appreciate the work you're doing as I can see it's taking a lot of effort.

Commit:
Allow a little more tolerance in this test on Cygwin

Commit:
A much cleaner, better functioning, and more Cygwin-ish approach to error
handling from Python wrappers for Windows APIs.

Instead of attempting to reimplement PyErr_SetFromWindowsError, on Cygwin we just
replace those calls with a macro wrapping PyErr_SetFromErrno which uses the
cygwin_internal API to map Windows error codes to POSIX error codes just in the same
way Cygwin would.

This appears to work correctly (for example trying to get info on a non-existent
process returns errno ESRCH (No such process).

Commit:
As noted in the comment, cygpid_to_winpid and winpid_to_cygpid can have
their pure Python implementations replaced with more efficient (and less
racy) C implementations.

Commit:
First move of a utility common to both the windows and cygwin modules.

Although I considered adding a new module for this purpose, for now it
suffices to use the _common module for this (albeit a bit hackishly).

In this case the C wrapper for this function makes sense to put in
arch/windows.

KSSea commented:
Just curious if there's been any progress with this.  I'm very interested in having psutil available under Cygwin.

joaoe commented on psutil/_common.py:
```diff
@@ -32,10 +32,14 @@
     enum = None
 
+
+PY3 = sys.version_info[0] == 3
```
I suggest `>=` else sometime in the future Python 4 is released and no one knows why psutil is broken in cygwin.

embray replied:
Agree; thanks for pointing that out.  In fact instead of a `PY3` variable it'd be better to use `PY2` for Python < 3.

embray commented:
@KSSea Yes, I actually have some updates to this that I haven't pushed yet. I think last I left off I had rebased and done some refactoring and fixed some of my TODO items, but some of the tests weren't passing, and then I just got side-tracked.

giampaolo commented:
It turns out that apparently there's quite some demand for cygwin support:
https://www.google.it/search?q=psutil+cygwin&oq=psutil+cygwin&aqs=chrome..69i57j69i60l2j69i61j0.1753j0j7&sourceid=chrome&ie=UTF-8

embray commented:
Ugh, rebasing this on the current master was quite a mess.  It's going to take me a bit to untangle.

yselkowitz commented:
Any chance to finish this port?  I just came across something that requires psutil, and can certainly help get it the Cygwin distribution.

embray commented:
In fact one of the main issues that's been slowing this down is that discussed in #1247.  I'm working on a separate branch to kind of decouple some of the C code from Python in a way that makes refactoring easier.

That said, I know a lot of people are asking for this, so maybe if I can finish just some basic refactoring first the rest we can worry about later.  At this stage the biggest impediment is just getting caught up with master:   I started this work more than a year ago, and when I first posted this PR it had all tests passing, but not anymore.  

Incidentally I was just working on this last week.  It's not abandoned, just, low on my plate and something I poke at every now and then.  Which means it gets further behind master, and so on.... :)

giampaolo commented:
I may be wrong but it appears to me the main and first goal here should be avoiding to duplicate code from `_psutil_windows.c`, aka do not have `_psutil_cygwin.c` at all. IMO, `setup.py` should just do this (aka compile `_psutil_windows.c` and `_psutil_posix.c`):

```diff
diff --git a/setup.py b/setup.py
index fffefd2..0d55eb6 100755
--- a/setup.py
+++ b/setup.py
@@ -29,6 +29,7 @@ HERE = os.path.abspath(os.path.dirname(__file__))
 sys.path.insert(0, os.path.join(HERE, "psutil"))
 
 from _common import BSD  # NOQA
+from _common import CYGWIN  # NOQA
 from _common import FREEBSD  # NOQA
 from _common import LINUX  # NOQA
 from _common import NETBSD  # NOQA
@@ -103,19 +104,32 @@ def silenced_output(stream_name):
         setattr(sys, stream_name, orig)
 
 
-if WINDOWS:
+if WINDOWS or CYGWIN:
     def get_winver():
-        maj, min = sys.getwindowsversion()[0:2]
-        return '0x0%s' % ((maj * 100) + min)
-
-    if sys.getwindowsversion()[0] < 6:
+        if WINDOWS:
+            maj, min = sys.getwindowsversion()[0:2]
+            return '0x0%s' % ((maj * 100) + min)
+        else:
+            import re
+            winver_re = re.compile(r'CYGWIN_NT-(?P<major>\d+)\.(?P<minor>\d+)')
+            verstr = os.uname()[0]
+            m = winver_re.search(verstr)
+            maj = int(m.group('major'))
+            min = int(m.group('minor'))
+            return '0x0%s' % ((maj * 100) + min)
+
+    if WINDOWS and sys.getwindowsversion()[0] < 6:
         msg = "warning: Windows versions < Vista are no longer supported or "
         msg = "maintained; latest official supported version is psutil 3.4.2; "
         msg += "psutil may still be installed from sources if you have "
         msg += "Visual Studio and may also (kind of) work though"
         warnings.warn(msg, UserWarning)
 
-    macros.append(("PSUTIL_WINDOWS", 1))
+    if WINDOWS:
+        macros.append(("PSUTIL_WINDOWS", 1))
+    else:
+        macros.append(("PSUTIL_CYGWIN", 1))
+
     macros.extend([
         # be nice to mingw, see:
         # http://www.mingw.org/wiki/Use_more_recent_defined_functions
```

How feasible would it be to restart from scratch and just compile `_psutil_windows.c` and `_psutil_posix.c`? With that in place I have a pretty clear idea on how to refactor the py code in order to reuse `_pslinux.py`.

embray commented:
I'm working on refactoring `_psutil_cygwin.c` so that there is less duplication.  However, it does not make sense to compile `_psutil_window.c` on Cygwin.  I don't think it will even build without significant `#ifdef`s, and even then much of the module is not applicable to Cygwin.

On the other hand, there is some Cygwin-specific code in `_psutil_cygwin.c` that really wouldn't make sense to have anywhere else, so it doesn't make sense to get rid of it entirely, with one exception: I have since released a package called pycygwin which includes basically the same Cygwin-specific functionality that I added in `_psutil_cygwin.c`.  If I could make that cygwin-only dependency of psutil then that code could be reused.  Otherwise it makes sense to copy it.  I should note that for pycygwin I used Cython, whereas the code in `_psutil_cygwin.c` of course just uses the C-API directly.

embray commented:
I think the biggest problem with this PR, and @giampaolo I think you'll agree, is that it was too ambitious.  I tried to get every, or almost every interface working just to see if I could.

But for the purposes of getting something that at least partially works as a first pass, I think I might start over again (although still reuse a lot) with my sights set a bit smaller.

The thing about Cygwin is that it is supposed to be *mostly* POSIX compatible.  There are a few POSIX APIs that are not implemented (except as stubs), and a few narrow areas where it is necessarily *not* POSIX compatible.  There are also a few non-POSIX interfaces borrowed from Linux.  So I believe that a lot of ground can be covered by just a little bit of refactoring to `_pslinux.py` (I would like to add something like a `_pslinux_common.py` that contains bits that can be used by other Linux-like platforms; alternatively it could just go in the existing `_common.py`).  Given that, I can add a simpler `_pscygwin.py` that only supports what is possible using bits from the posix and linux modules.

In general one should *not* use the Windows APIs on Cygwin; the whole point is that it's supposed to abstract away the Windows API.  But of course you *can* do it, and as my prototype demonstrates you can still add a lot of functionality if you dig in.    In some small cases it might still be unavoidable.  Though I also think a lot of code from psutil, as long as the license allows it, could form the basis for a new features in Cygwin to implement a lot of currently missing interfaces (I'm especially thinking `/proc/net` in this case).

giampaolo commented:
So, here's some comments in light of the recent discussion occurred in #1342. As I said in there the main goal here should be to avoid code duplication. Since you're going to restart from scratch I propose to proceed step by step. The goal of the first step IMO should be solving the import time issues and get to a point where `import psutil` just works. That will help us have a clearer view on how to proceed next. As part of the first step I suggest to leave the C alone and just focus on the python/import part. I suggest the following:

1) define a `psutil/_pscygwin.py` which will include the Python logic that differs from Linux (no Linux logic should be copied in here)
2) on Cygwin `psutil/__init__.py` will import `psutil/_pslinux.py`: 

```python
# __init__.py
if CYGWIN:
    from . import _pslinux as _psplatform
```

3) in `psutil/_pslinux.py` you can work with `if CYGWIN: ...` in order to exclude linux-specific logic which is loaded on import. More or less:

```python
# _pslinux.py

if CYGWIN:
    import _pscygwin
else:
    from . import _psutil_linux as cext

...

HAS_PRLIMIT = not CYGWIN and hasattr(cext, "linux_prlimit")
if not CYGWIN:
    try:
        set_scputimes_ntuple("/proc")
    except Exception:
        # Don't want to crash at import time.
        traceback.print_exc()
        scputimes = namedtuple('scputimes', 'user system idle')(0.0, 0.0, 0.0)

...

# at the end of the module, replace the APIs which differ
class Process:   # <-- the linux process
    ...

if CYGWIN:
    class Process(_pscygwin.Process):
        pass

virtual_memory = _pscygwin.virtual_memory
net_io_counters = _pscygwin.net_io_counters
...
```

* `psutil/_pscygwin.py` will more or less look like this

```python
import _psutil_cygwin as cext

def virtual_memory:
    ...

def net_io_counters():
    ...

class Process:
    def threads(self):
        return cext.proc_threads(self.pid)
...
```

The impact on `_pslinux.py` is affordable because you'll only change a couple of things which are loaded at import time. The rest will remain unchanged and clearly defined as it will keep living in the body functions/methods. These parts will never be invoked on Cygwin because you'll overwrite `_pslinux.py` namespace by injecting the definitions coming from `_pscygwin.py` at runtime. 

This IMO is a reasonable approach as long as we're not clear how the different bits should be moved around and refactored. 

Once you're done with that the next logical step will be identifying what C functions from `_psutil_windows.c`. That is more sensitive and less clear to me. We may proceed by reusing `_psutil_windows.c` and add `#ifdef`s all around, or we may have a separate `_psutil_cygwin.c` file. But that's not clear at this point (to me at least) so let's leave it for later and let's set the milestone for step 1 only.

@embray hope this helps and let me know if the above makes sense to you.

giampaolo commented:
Actually, on a second though I think it's better to do the other way around: to have `_pscygwin.py` import `_pslinux.py`:

`psutil/__init__.py`:

```python
if CYGWIN: 
    from . import _pscygwin as _psplatform
```

`psutil/_pslinux.py` (just solve the import time issues):

```python
if not CYGWIN:
    from . import _psutil_linux as cext

HAS_PRLIMIT = not CYGWIN and hasattr(cext, "linux_prlimit")
if not CYGWIN:
    try:
        set_scputimes_ntuple("/proc")
    except Exception:
        # Don't want to crash at import time.
        traceback.print_exc()
        scputimes = namedtuple('scputimes', 'user system idle')(0.0, 0.0, 0.0)
```

`psutil/_pscygwin.py` (redefine what's needed):

```python
from . import _pslinux
from . import _psutil_posix as cext_posix

net_if_addrs = cext_posix.net_if_addrs
disk_usage = _psposix.disk_usage

class Process(_pslinux.Process):

    def threads(self):
        return cext.proc_threads(self.pid)
```

embray commented:
@giampaolo I'm sorry you took the time to write all this, because I'm way ahead of you on most of that.  I have a different approach that I think you'll appreciate that was designed to avoid littering the code with `if (not) CYGWIN` blocks everywhere, which ends up looking quite a bit cleaner, for now.

I don't believe it's good to use `_pslinux.py` at all in the Cygwin implementation as there are too many differences between the platforms (Cygwin borrows from Linux in some areas that it likes, such as the layout of the `/proc` filesystem, but it is *not* Linux).  You'll just wind up with tons of the code in `_pslinux.py` bracketed by `if not CYGWIN`.  No good.

There is still a *tiny* bit of duplication for now, but not much, and it can be avoided with some other refactoring ideas I have.

I'm just making sure that the applicable tests pass, and that all other tests for interfaces that are not implemented yet are skipped as appropriate. Please in the meantime don't give it further thought until I post some actual code.

giampaolo commented:
You may have misunderstood what I have in mind. What I propose only adds a couple of `if (not) CYGWIN` in `_pslinux.py` in order to avoid crashing at import time, but all the rest is going to live in `_pscygwin.py` and reused via aliasing or subclassing. As such I don't aim at littering `_pslinux.py` body functions with `if (not) CYGWIN` (I for one would be against it). Judging from what I see in the current PR there are different things in common with Linux:

```
cpu_count_logical()
cpu_count_physical()
disk_partitions()
pids()
Process.name()
Process.terminal()
Process.cpu_times()
Process.wait()
Process.cpu_affinity()
Process.ppid()
```
I am not sure what you have in mind but FYI I am not keen on moving those parts out of `_pslinux.py` just for reusing them in `_pscygwin.py`. Linux is the major platform here (and in psutil in general), so it should not be altered that much just to integrate Cygwin, which really is a minor one by comparison.

giampaolo commented:
@embray after re reading your diff I noticed there are some tiny differences. I missed that on my first run looking back at this PR. I read your last message without giving it the proper attention and I was also a bit harsh in dismissing it. I am sorry about that.

embray commented:
No problem.  I'm still working on getting the tests in a sane state, but to spare some suspense what I'm doing is I created a `_pslinux_common.py` module, moved some of the common functionality there (there's not as much yet as you might assume), and both `_pscygwin.py` and `_pslinux.py` use it.

giampaolo commented:
Thanks for the update @embray. OK, this time I will shut up and wait to see some code and then we'll elaborate on that. FWIW, I Googled a bit and it seems Cygwin is not so "minor" as I thought. It appears there's quite some demand for it so I'm all for getting this in as a major feature. Should we close this PR?

embray commented:
Still making progress on this.  I've had various other fires to put out but I'm close to having the new prototype ready.  It's been mostly a progress of adding `skipIfs` to the right tests, as my current approach is to support just bare minimum functionality on Cygwin to start with so that we at least have a stub on which to add additional functionality (much of which can still be preserved from this PR, but added a bit at a time).

giampaolo commented:
Closing this out as outdated.
